x-pack/plugins/endpoint/public/applications/endpoint/store/policy_list/test_mock_utils.ts

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { HttpStart } from 'kibana/public';
+import { GetDatasourcesResponse, INGEST_API_DATASOURCES } from '../../services/ingest';
+import { EndpointDocGenerator } from '../../../../../common/generate_data';
+import { AppAction, GlobalState, MiddlewareFactory } from '../../types';
+
+const generator = new EndpointDocGenerator('policy-list');
+
+/**
+ * It sets the mock implementation on the necessary http methods to support the policy list view
+ * @param mockedHttpService
+ * @param responseItems
+ */
+export const setPolicyListApiMockImplementation = (
+  mockedHttpService: jest.Mocked<HttpStart>,
+  responseItems: GetDatasourcesResponse['items'] = [generator.generatePolicyDatasource()]
+): void => {
+  mockedHttpService.get.mockImplementation((...args) => {
+    const [path] = args;
+    if (typeof path === 'string') {
+      if (path === INGEST_API_DATASOURCES) {
+        return Promise.resolve<GetDatasourcesResponse>({
+          items: responseItems,
+          total: 10,
+          page: 1,
+          perPage: 10,
+          success: true,
+        });
+      }
+    }
+    return Promise.reject(new Error(`MOCK: unknown policy list api: ${path}`));
+  });
+};
+
+/**
+ * Utilities for testing Redux middleware
+ */
+export interface MiddlewareActionSpyHelper<S = GlobalState> {
+  /**
+   * Returns a promise that is fulfilled when the given action is dispatched or a timeout occurs.
+   * The use of this method instead of a `sleep()` type of delay should avoid test case instability
+   * especially when run in a CI environment.
+   *
+   * @param actionType
+   */
+  waitForAction: (actionType: Pick<AppAction, 'type'>['type']) => Promise<void>;
+  /**
+   * Redux middleware that enables spying on the action that are dispatched through the store
+   */
+  actionSpyMiddleware: ReturnType<MiddlewareFactory<S>>;
+}
+
+/**
+ * Creates a new instance of middleware action helpers
+ * Note: in most cases (testing concern specific middleware) this function should be given
+ * the state type definition, else, the global state will be used.
+ *
+ * @example
+ * // Use in Policy List middleware testing
+ * const { actionSpyMiddleware, waitForAction } = createSpyMiddleware<PolicyListState>();
+ * store = createStore(
+ *    policyListReducer,
+ *    applyMiddleware(
+ *      policyListMiddlewareFactory(fakeCoreStart, depsStart),
+ *      actionSpyMiddleware
+ *    )
+ *  );
+ * // later in test
+ * it('...', async () => {
+ *   //...
+ *   await waitForAction('serverReturnedPolicyListData');
+ *   // do assertion
+ * });
+ */
+export const createSpyMiddleware = <S = GlobalState>(): MiddlewareActionSpyHelper<S> => {
```

Comment:
@parkiino , @kevinlog , @oatkiller , @peluja1012 , @dplumlee , @bkimmel , @kqualters-elastic - FYI. If you find that you have the same need for this type of utility while creating test cases for middleware or views that trigger middleware, we can elevate this utility to be in a higher level folder and re-use it. Its currently built in a generic way, so it should adopt to other middleware testing cases.

@oatkiller  - I'm going to try to `jest.fn()` on the middleware idea next. Did not include it here because I did not have a need currently for it. 😃

Reply:
FYI - I made the middleware `dispatch` function a jest mocked function and exposed its `mock` object via the `dispatchSpy` property in the returned object.

Reply:
this is greatttt